compiler/rustc_privacy/src/lib.rs

```diff
@@ -2080,11 +2058,12 @@ fn check_private_in_public(tcx: TyCtxt<'_>, (): ()) {
     };
     tcx.hir().walk_toplevel_module(&mut visitor);
 
-    let has_pub_restricted = {
-        let mut pub_restricted_visitor = PubRestrictedVisitor { tcx, has_pub_restricted: false };
-        tcx.hir().walk_toplevel_module(&mut pub_restricted_visitor);
-        pub_restricted_visitor.has_pub_restricted
-    };
+    let has_pub_restricted = tcx.resolutions(()).visibilities.iter().any(|(&def_id, &v)| match v {
+        ty::Visibility::Public | ty::Visibility::Invisible => false,
+        ty::Visibility::Restricted(module) => {
+            module != tcx.parent_module_from_def_id(def_id).to_def_id()
```

Comment:
This is not exactly the same thing because e.g. `pub(self)` will be treated differently, but it doesn't matter, it's a hack and heuristic anyway.

Reply:
I fear this heuristic is responsible for the perf regression: many more calls to `parent_module_from_def_id`.
What is the point of this heuristic?

Reply:
`rustc_privacy` has two private-in-public checkers
- old and very buggy `ObsoleteVisiblePrivateTypesVisitor`
- properly implemented `PrivateItemsInPublicInterfacesVisitor`

The latter one caused too much breakage when originally implemented (https://github.com/rust-lang/rust/issues/34537), so it was decided to report errors from the new checker as warnings when the old checker does not report them.

"Restricted" visibilities (`pub(...)`) were unstable back then, so we could break code using them, so the new checker was enabled for such code without any `ObsoleteVisiblePrivateTypesVisitor`-based compatibility measures.

Reply:
All these hacks will become obsolete once https://github.com/rust-lang/rust/issues/48054 is implemented.

For now I suggest to just evaluate the `has_pub_restricted` during ~AST lowering~ late resolution and put in into ~HIR~ resolver outputs.